File: src/main/java/com/mooregreatsoftware/gradle/defaults/config/BintrayConfig.java
/*
 * Copyright 2014-2016 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooregreatsoftware.gradle.defaults.config;

import com.google.common.io.Files;
import com.jfrog.bintray.gradle.BintrayExtension;
import com.jfrog.bintray.gradle.BintrayPlugin;
import com.mooregreatsoftware.gradle.defaults.DefaultsExtension;
import lombok.val;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.gradle.api.Project;

import java.io.File;
import java.util.HashMap;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static java.util.Optional.ofNullable;

@SuppressWarnings({"WeakerAccess", "RedundantCast", "RedundantTypeArguments"})
public class BintrayConfig extends AbstractConfigWithExtension {
    public BintrayConfig(Project project, DefaultsExtension extension) {
        super(project, extension);
    }


    public void config() {
        plugins().withType(BintrayPlugin.class, plugin -> {
            if (project.hasProperty("bintrayUser") && project.hasProperty("bintrayKey")) {
                project.afterEvaluate(prj -> configBintray());
            }
        });
    }


    private void configBintray() {
        val bintray = bintrayExtension(project);

        bintray.setUser((String)project.property("bintrayUser"));
        bintray.setKey((String)project.property("bintrayKey"));

        bintray.setPublications("main");
        bintray.setPublish(true);

        val pkgConfig = bintray.getPkg();

        if (extension.getOrgName() != null) pkgConfig.setUserOrg(extension.getId());

        ofNullable(extension.getBintrayRepo()).ifPresent(pkgConfig::setRepo);
        ofNullable(extension.getBintrayPkg()).ifPresent(pkgConfig::setName);
        ofNullable(project.getDescription()).ifPresent(pkgConfig::setDesc);
        ofNullable(extension.getSiteUrl()).ifPresent(pkgConfig::setWebsiteUrl);
        ofNullable(extension.getIssuesUrl()).ifPresent(pkgConfig::setIssueTrackerUrl);
        ofNullable(extension.getVcsReadUrl()).ifPresent(pkgConfig::setVcsUrl);
        ofNullable(extension.getLicenseKey()).ifPresent(pkgConfig::setLicenses);

        val bintrayLabels = toStringArray(extension.getBintrayLabels());
        pkgConfig.setLabels(bintrayLabels);
        pkgConfig.setPublicDownloadNumbers(true);

        val bintrayPkgVersion = pkgConfig.getVersion();

        bintrayPkgVersion.setVcsTag("v" + project.getVersion());
        bintrayPkgVersion.setAttributes(bintrayAttributes());

        val gpg = bintrayPkgVersion.getGpg();
        gpg.setSign(true);
        if (project.hasProperty("gpgPassphrase")) {
            gpg.setPassphrase((String)project.property("gpgPassphrase"));
        }
    }


    private static String[] toStringArray(@Nullable Set<String> labels) {
        return (@NonNull String @NonNull [])ofNullable(labels).
            map(set -> set.toArray(new String[set.size()])).
            orElse(new String[0]);
    }


    private HashMap<String, Object> bintrayAttributes() {
        val bintrayAttributes = new HashMap<String, Object>();
        if (ofNullable(extension.getBintrayLabels()).filter(lbl -> lbl.contains("gradle")).isPresent()) {
            info("bintrayLabels does includes 'gradle' so generating 'gradle-plugins' attribute");
            val filesTree = gradlePluginPropertyFiles(project);
            val pluginIdBintrayAttributeValues = filesToPluginIds(filesTree).
                map(this::pluginIdToBintrayAttributeValue).
                collect(Collectors.<@NonNull String>toList());
            bintrayAttributes.put("gradle-plugins", pluginIdBintrayAttributeValues);
        }
        else {
            info("bintrayLabels does not include 'gradle' so not generating 'gradle-plugins' attribute");
        }

        info("bintrayAttributes: " + bintrayAttributes);
        return bintrayAttributes;
    }


    private String pluginIdToBintrayAttributeValue(String pluginId) {
        return pluginId + ":" + project.getGroup() + ":" + name();
    }


    private static Stream<String> filesToPluginIds(Iterable<File> filesTree) {
        return StreamSupport.stream(filesTree.spliterator(), false).
            map(BintrayConfig::filenameNoExtension);
    }


    private static String filenameNoExtension(File file) {
        return Files.getNameWithoutExtension(file.getName());
    }


    private static Iterable<File> gradlePluginPropertyFiles(Project project) {
        val fileTreeConf = new HashMap<String, String>();
        fileTreeConf.put("dir", "src/main/resources/META-INF/gradle-plugins");
        fileTreeConf.put("include", "*.properties");
        return project.fileTree(fileTreeConf);
    }


    public static BintrayExtension bintrayExtension(Project project) {
        return project.getConvention().getByType(BintrayExtension.class);
    }

}

File: src/main/kotlin/com/mooregreatsoftware/gradle/defaults/config/BintrayConfig.kt
/*
 * Copyright 2014-2016 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooregreatsoftware.gradle.defaults.config

import com.google.common.io.Files
import com.jfrog.bintray.gradle.BintrayExtension
import com.jfrog.bintray.gradle.BintrayPlugin
import com.mooregreatsoftware.gradle.defaults.ReadableDefaultsExtension
import com.mooregreatsoftware.gradle.defaults.Ternary
import com.mooregreatsoftware.gradle.defaults.postEvalCreate
import com.mooregreatsoftware.gradle.defaults.readableDefaultsExtension
import org.gradle.api.GradleException
import org.gradle.api.Project
import java.io.File
import java.util.HashMap
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit

class BintrayConfig private constructor(private val project: Project, readExtensionFuture: Future<ReadableDefaultsExtension>) {

    init {
        project.postEvalCreate {
            val readExtension = readExtensionFuture.get(20, TimeUnit.SECONDS)
            val bintray = project.bintrayExtension()!!

            project.gradle.taskGraph.addTaskExecutionGraphListener {
                if (it.hasTask("bintrayUpload")) {
                    if (project.hasProperty("bintrayUser") && project.hasProperty("bintrayKey")) {
                        bintray.user = project.property("bintrayUser") as String
                        bintray.key = project.property("bintrayKey") as String
                    }
                    else {
                        throw GradleException("You need to set the \"bintrayUser\" and \"bintrayKey\" properties on the project to upload to BinTray")
                    }
                }
            }

            bintray.setPublications("main")
            bintray.isPublish = true

            val pkgConfig = bintray.pkg

            if (readExtension.orgName != null) pkgConfig.userOrg = readExtension.orgId

            pkgConfig.setPackageRepo(readExtension)
            pkgConfig.setPackageName(readExtension)

            project.description?.let { pkgConfig.desc = it }
            pkgConfig.websiteUrl = readExtension.siteUrl
            pkgConfig.issueTrackerUrl = readExtension.issuesUrl
            pkgConfig.vcsUrl = readExtension.vcsReadUrl
            pkgConfig.setLicenses(readExtension.licenseKey)

            val bintrayLabels = toStringArray(readExtension.bintrayLabels)
            pkgConfig.setLabels(*bintrayLabels)
            pkgConfig.isPublicDownloadNumbers = true

            val bintrayPkgVersion = pkgConfig.version

            bintrayPkgVersion.vcsTag = "v" + project.version
            bintrayPkgVersion.attributes = bintrayAttributes(readExtension)

            val gpg = bintrayPkgVersion.gpg
            if (project.hasProperty("gpgPassphrase")) {
                gpg.sign = true
                gpg.passphrase = project.property("gpgPassphrase") as String
            }
            else {
                project.logger.info("\"gpgPassphrase\" not set on the project, so not signing the Bintray upload")
            }
        }
    }

    private fun BintrayExtension.PackageConfig.setPackageRepo(readExtension: ReadableDefaultsExtension) {
        if (this.repo.isNullOrBlank()) {
            if (readExtension.bintrayRepo.isNullOrBlank())
                throw GradleException("Need to set defaults { bintrayRepo = ... }")
            this.repo = readExtension.bintrayRepo
        }
        else {
            if (!readExtension.bintrayRepo.isNullOrBlank())
                throw GradleException("Both defaults { bintrayRepo = ... } and bintray { pkg { repo = ... } } have been set.")
            // repo has already been set
        }
    }

    private fun BintrayExtension.PackageConfig.setPackageName(readExtension: ReadableDefaultsExtension) {
        if (this.name.isNullOrBlank()) {
            this.name = when {
                readExtension.bintrayPkg.isNullOrBlank() -> project.name
                else -> readExtension.bintrayPkg
            }
        }
        else {
            if (!readExtension.bintrayPkg.isNullOrBlank())
                throw GradleException("Both defaults { bintrayPkg = ... } and bintray { pkg { name = ... } } have been set.")
            // name has already been set
        }
    }


    private fun bintrayAttributes(readExtension: ReadableDefaultsExtension): Map<String, Any> {
        val bintrayAttributes = HashMap<String, Any>()
        if (readExtension.bintrayLabels?.contains("gradle") ?: false) {
            project.logger.info("bintrayLabels does includes 'gradle' so generating 'gradle-plugins' attribute")
            val filesTree = gradlePluginPropertyFiles(project)
            val pluginIdBintrayAttributeValues = filesToPluginIds(filesTree).map { this.pluginIdToBintrayAttributeValue(it) }.toList()
            bintrayAttributes.put("gradle-plugins", pluginIdBintrayAttributeValues)
        }
        else {
            project.logger.info("bintrayLabels does not include 'gradle' so not generating 'gradle-plugins' attribute")
        }

        project.logger.info("bintrayAttributes: $bintrayAttributes")
        return bintrayAttributes
    }


    private fun pluginIdToBintrayAttributeValue(pluginId: String) = "$pluginId:${project.group}:${project.name}"

    companion object {

        private fun create(project: Project): Future<BintrayConfig?> {
            return confFuture(project, "BinTray",
                { if (project.plugins.hasPlugin(BintrayPlugin::class.java)) Ternary.TRUE else Ternary.MAYBE },
                { project.plugins.hasPlugin(BintrayPlugin::class.java) },
                { BintrayConfig(project, project.readableDefaultsExtension()) },
                null
            )
        }

        /**
         * Returns the BintrayConfig for in the given Project.
         *
         * @param project the project containing the BintrayConfig
         */
        @JvmStatic fun of(project: Project): Future<BintrayConfig?> {
            return ofFuture(project, { create(project) })
        }

    }
}

private fun toStringArray(labels: Set<String>?): Array<String> = when (labels) {
    null -> emptyArray()
    else -> labels.toTypedArray()
}


private fun filesToPluginIds(filesTree: Iterable<File>) = filesTree.map { filenameNoExtension(it) }


private fun filenameNoExtension(file: File) = Files.getNameWithoutExtension(file.name)


private fun gradlePluginPropertyFiles(project: Project): Iterable<File> = project.fileTree(
    mapOf("dir" to "src/main/resources/META-INF/gradle-plugins", "include" to "*.properties")
)


fun Project.bintrayExtension(): BintrayExtension? = this.convention.findByType(BintrayExtension::class.java)
